Conditionally set output_paths based on Remote Executor capabilities

sluongng opened the pull request:
This is a follow up to #18269, toward the discussion in #18202.

Bump the Remote API supported version to v2.1.

Based on the Capability of the Remote Executor, either use output_paths
field or the legacy fields output_files and output_directories.

sluongng commented:
It might be interesting for subsequent patches to add the ability to set RemoteWorker's ApiVersion via a flag.
That should enable us to write more integration tests for different version behaviors.

However, that's out of the scope of the current PR.

sluongng commented:
cc: @coeuvre @tjgq @EdSchouten @fmeum 

This is probably the biggest and final change in my PR chain about output_paths and output_symlinks.
Do you guys think this is backward compatible enough that it could go into 6.3.0 or do we have to wait for 7.0.0.

sluongng commented:
ping @coeuvre @tjgq I would appreciate getting some feedback here. 

This PR has been out for a month now. The longer it's not merged, the more effort is required to rebase it against conflicts in master.

coeuvre reviewed (approved):
Sorry for the delay. LGTM!

coeuvre commented:
@bazel-io fork

coeuvre commented:
@bazel-io fork 6.3.0

coeuvre commented:
This change breaks remote server that only support REAPI v2.0 (because the change of `current`). It is bing rolled back.

sluongng commented:
@coeuvre any advice on how I should fix the issue?

I am not sure how `current` is being used as we do alter the implementation based on Remote Exec server capabilities.
All the tests also pass using RBE service on CI, which I assumed are 2.0 only?

coeuvre commented:
The tests are using [remote worker](https://cs.opensource.google/bazel/bazel/+/master:src/tools/remote/src/main/java/com/google/devtools/build/remote/worker/RemoteWorker.java) instead RBE so we didnt' catch the regression.

`current` is used to check against remote server's [low and high version](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/RemoteServerCapabilities.java;l=186;drc=0cdd40972ba5253b4b4f0de80f080a911dc8c99b). Simply increasing it to 2.1 will fail the check for server whose high is 2.0.

Ideally, Bazel should also has the concept of low and high version, and compare it against server's low and high version. If there is an intersection, use highest version in that range - maybe call it resolved version. All the feature test later in the code path should then depend on the resolved version.

sluongng commented:
Let me see if I could send a separate PR to make Bazel support a high and low API version.

Thanks for the pointer!

sluongng commented:
This PR is currently pending on https://github.com/bazelbuild/bazel/pull/18647 changes

sluongng commented:
I rebased this on top of latest master.

Added a `--legacy_api` flag to `worker` binary. This allow me to set up 2 shell tests validating that Bazel would still work with remote cache + remote executor using version 2.0.

Some files diff are a bit noisy due to formatting. If that's a problem, please let me know and I could revert those hunks.

cc: @coeuvre @tjgq

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:
```diff
@@ -543,8 +550,20 @@ public RemoteAction buildRemoteAction(Spawn spawn, SpawnExecutionContext context
       }
 
+      var useOutputPaths = true;
+      if (mayBeExecutedRemotely(spawn)) {
```
What's the concern for not checking the server capabilities for cache-only actions?

My understanding is we should use the same message format for both remote cache and remote execution.

coeuvre replied:
Or maybe we should remove the check. The problem now is we only store the server capabilities in `remoteExecutor` which could be `null` in case of remote cache only build. Can we store the server capabilities in `RemoteExecutionService` instead?

sluongng replied:
I think we could switch to storing ServerCapabilities inside RemoteCache instead of CacheCapabilities.
That way we could access it here using the logic I proposed above, like this

```java
  private boolean shouldUseOutputPaths(Spawn spawn) {
    var isRemoteExec = mayBeExecutedRemotely(spawn);
    var isRemoteCached = mayBeCachedRemotely(spawn);
    if (!isRemoteExec || !isRemoteCached) {
      return true;
    }

    ServerCapabilities serverCapabilities;
    if (isRemoteExec) {
      serverCapabilities = remoteExecutor.getServerCapabilities();
    } else {
      serverCapabilities = remoteCache.getServerCapabilities();
    }
    if (serverCapabilities == null) {
      return true;
    }

    var supportStatus = ClientApiVersion.current.checkServerSupportedVersions(serverCapabilities);
    if (supportStatus.isUnsupported()) {
      return true;
    }

    return supportStatus.getHighestSupportedVersion().compareTo(ApiVersion.twoPointOne) >= 0;
  }
```

This solution should work with the exception to the case where the RemoteCache and the RemoteExecutor support 2 different low/high version pairs in their capability. Using the logic above ☝️  we would bias to using Executor capabilities instead of RemoteCache capabilities. But a future change could be made so that we could use the lowest supported version among the 2.

WDYT? I could send out a separate PR for the RemoteCache refactoring to `s/CacheCapabilities/ServerCapabilities/g`, or include that as part of this PR as a separate commit.

coeuvre replied:
That could also work. Please create a separate PR for the RemoteCache refactoring. Thanks!

sluongng replied:
I sent https://github.com/bazelbuild/bazel/pull/18734 for that.

I also rebase this PR on top of that PR and included https://github.com/bazelbuild/bazel/pull/18270/commits/c5b59e65ed4f2256c2ea7b8a9da34757f068ee23 to demonstrate how the end result would look like.

sluongng replied:
@coeuvre Thanks for this pointer. I forgot about this thread.

I went back to #18734 and found it mostly no longer relevant because of the recent refactoring in master branch. I have added the needed API as a separate commit here though.

This makes the current PR look like this
```
> git log --oneline -6 --reverse
c08353126d RemoteCacheClient: add getServerCapabilities
afce68f4b8 Apply google-java-format to RemoteWorkerOptions
11c09c134f Conditionally set output_paths based on Remote Executor capabilities
c43cd180a8 init useoutputpaths
2de7c2d8f9 RemoteOptions: add useOutputPaths
2d7f14fb6a RemoteOptions: flip remote_use_output_paths
```

The first 5 commits could be backported to 7.3.0 if needed. The 6th commit, which flip the flag value, could be an 8.0-only change.
I have tested the first 5 commits exclusively to make sure they pass CI. Though, I have to admit, the additional flag made testing a bit sparse. I hope it's short-lived enough for that to not matter.

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:
```diff
@@ -522,6 +534,31 @@ private ToolSignature getToolSignature(Spawn spawn, SpawnExecutionContext contex
 
+  private boolean shouldUseOutputPaths(Spawn spawn) {
+    var isRemoteExec = mayBeExecutedRemotely(spawn);
+    var isRemoteCached = mayBeCachedRemotely(spawn);
```
This doesn't work for combined cache. And what happens if remote cache and remote executor return different API version?

I don't get why we need two server capabilities. It seem likes it only makes thing more complicated. Bazel should use the same API version for both remote cache and remote execution (if the two endpoints are different).

sluongng replied:
> I don't get why we need two server capabilities. It seem likes it only makes thing more complicated. Bazel should use the same API version for both remote cache and remote execution (if the two endpoints are different).

So in the case of remote cache and remote executor differ, we know that we could have 2 server capabilities regardless.

When that happens, we _SHOULD_ run some validation against the 2 server capabilities to figure out what is the API version range. And in case the version ranges of Remote Cache and Remote Executor do not overlap, then we err out.

With that in mind, I will attempt to implement that logic in RemoteModule. Then in RemoteExecutionService, we should be able to use one server capabilities instead of 2.

coeuvre replied:
SGTM!

EdSchouten commented on src/main/java/com/google/devtools/build/lib/remote/ApiVersion.java:
```diff
@@ -23,12 +23,18 @@ public class ApiVersion implements Comparable<ApiVersion> {
-  public static final ApiVersion high =
-      new ApiVersion(SemVer.newBuilder().setMajor(2).setMinor(0).build());
+  public static final ApiVersion low = twoPointZero;
+  public static final ApiVersion high = twoPointOne;
```
I don't think this is accurate. With this change applied, Bazel supports up to and including 2.3 just fine:

- We already send platform properties as part of the Action message, as required by REv2.2.
- We already rely on `argv[0]` to be resolved relative to the working directory, using `$PATH` lookups, as required by REv2.3.

So in that sense Bazel will behave more like an REv2.3 client than REv2.1 with this change applied.

sluongng replied:
My plan was to finish this PR with 2.1 first, then a subsequent PR could raise it from 2.1 -> 2.3.
That would allow the change to be more atomic if things need to be reverted (which did happen).

I have been busy with more customer-facing tasks so this PR has been slow to iterate upon. 🐌
I hope I could spend a bit more time on this PR at the end of this week / early next week. 🤞

sluongng replied:
Bumped to 2.2 since 2.3 tag in remote-apis I not yet created https://github.com/bazelbuild/remote-apis/tags

sluongng commented:
@coeuvre sorry for the delay on this PR. I finally get some time to revisit my Bazel works. I have cherry-picked + rebased the old PR on top of the latest origin/master.

The PR is still aiming to switch to using `Command.output_paths` and `ActionResult.output_symlinks` fields when the server capabilities indicate 2.1 is supported.

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:
```diff
@@ -585,8 +592,20 @@ public RemoteAction buildRemoteAction(Spawn spawn, SpawnExecutionContext context
         platform = PlatformUtils.getPlatformProto(spawn, remoteOptions);
       }
 
+      var useOutputPaths = true;
```
To be on the safe side, can we add a flag so we don't accidentally break people? We can flip the flag in 8 and remove it in 9.

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/options/RemoteOptions.java:
```diff
@@ -420,6 +420,16 @@ public RemoteBuildEventUploadModeConverter() {
   public boolean cacheCompression;
 
+  @Option(
+      name = "remote_use_output_paths",
```
Since this is more like a guard for the incompatible change than a feature, can we rename it to `--incompatible_remote_use_output_paths`? It's likely we will remove this flag in near future.

sluongng replied:
done

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/options/RemoteOptions.java:
```diff
@@ -420,6 +420,16 @@ public RemoteBuildEventUploadModeConverter() {
 
+  @Option(
+      name = "remote_use_output_paths",
+      defaultValue = "true",
```
Can you open a separate PR for the flag flip commit? When the PR is imported, all the individual commits are merged into one.

sluongng replied:
will do

sluongng replied:
https://github.com/bazelbuild/bazel/pull/23582

coeuvre reviewed (changes requested):
Thanks for the update! I left some comments but otherwise LGTM.

sluongng commented on src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:
```diff
@@ -564,6 +573,69 @@ private void maybeReleaseRemoteActionBuildingSemaphore() {
+      return;
+    }
+    // Default value
+    this.useOutputPaths = true;
```
done

sluongng commented on src/main/java/com/google/devtools/build/lib/remote/common/RemoteCacheClient.java:
```diff
@@ -35,6 +36,8 @@
 public interface RemoteCacheClient extends MissingDigestsFinder {
   CacheCapabilities getCacheCapabilities() throws IOException;
 
+  ServerCapabilities getServerCapabilities() throws IOException;
```
This could potentially be a large refactoring, so I would prefer to leave it to a separate followup PR instead. Let me know what you think

sluongng replied:
created https://github.com/bazelbuild/bazel/pull/23593 as a follow up

sluongng commented:
@coeuvre sorry for the slow turnaround. I have rebased the PR on top of the latest master and resolved all the merge conflicts as well as addressed the review comment.

Let me know what you think.

sluongng commented:
@coeuvre This finally passed CI. You might want to give this another read as the rebase came with some adjustments for the latest changes in `master` branch. 🙇

coeuvre reviewed (approved):
LGTM! Can you rebase again? I will do the import then.

sluongng commented:
Rebased. PTAL 🙏

coeuvre reviewed (approved):
Thanks! Importing.

fmeum commented:
@bazel-io fork 7.4.0
